src/ui/controllers/course/index.ts

```diff
@@ -99,8 +99,9 @@ export async function displayModule(
 			res.redirect(`/book/${course.id}/${module.id}/choose-date`)
 			break
 		case 'link':
+		case 'file':
 			await xapi.record(req, course, xapi.Verb.Experienced, undefined, module)
-			res.redirect(module.location!)
+			res.redirect(module.location! || module.url!)
```

Comment:
This looks wrong - but I've been doing some swift so might be getting confused. Force unwrapping module.location will throw an error before the or has a chance to check?

Reply:
Swift is what typescript wishes it was. Anyway, .location is "optional", so unwrap and check if its there, if it's undefined, try module.url

Reply:
although, if both fails, who knows what'll happen 😛